### PR TITLE
improve Array#push

### PR DESCRIPTION
**Pull request by MaxGraey**
- improve detecting reallocation condition

### CODE REVIEW

File: std/assembly/array.ts [hunk 1]
```diff
@@ -167,9 +167,9 @@ export class Array<T> {
     var buffer = this.buffer_;
     var capacity = buffer.byteLength >>> alignof<T>();
     var newLength = length + 1; // safe only if length is checked
-    if (<u32>length >= <u32>capacity) {
+    if (<u32>newLength >= <u32>capacity) {
```
Comment (dcodeIO) [inline std/assembly/array.ts]: If `length` isn't validated explicitly, checking `newLength` might use an overflown value. Once if `length` is known to be valid, `newLength` is known to be in range.

Comment (MaxGraey) [reply std/assembly/array.ts]: so it was legit before?

Comment (dcodeIO) [reply std/assembly/array.ts]: Don't know, what actual issue are you trying to fix?

Comment (MaxGraey) [reply std/assembly/array.ts]: It works in both cases, but modified version more clear in my opinion. And produce more optimal [output](https://github.com/AssemblyScript/assemblyscript/pull/300/files/84d396c4116ac43eb3bd744926da5159aa2b69f1#diff-d48956ee38e0493bf00054580270b5d3R3070)

**Comment by dcodeIO** [comment]
With the improvement this is still not correct because `newLength = length + 1` can overflow, if `length = 0xffffffff` is specified.

**Comment by MaxGraey** [comment]
But this not possible because `length` can't be greater than `MAX_LENGTH << alignof<T>()` which is `0x7FFFFFFF - HEADER_SIZE` according to [this](https://github.com/AssemblyScript/assemblyscript/blob/master/std/assembly/array.ts#L37) and this [this](https://github.com/AssemblyScript/assemblyscript/blob/master/std/assembly/array.ts#L59)

**Comment by dcodeIO** [comment]
Ah, yeah, it's reading `length` from the structure. Had in mind it's a parameter. So the issue I see is rather that the new check relies on assumptions, while the old check didn't have to because it checked the base value instead of a derived one.

**Pull request closed without merging**